src/Http/Http/src/BufferSegment.cs

```diff
@@ -46,25 +46,62 @@ public BufferSegment NextSegment
             }
         }
 
+        public void SetMemory(object memoryOwner)
+        {
+            if (memoryOwner is IMemoryOwner<byte> owner)
+            {
+                SetMemory(owner);
+            }
+            else if (memoryOwner is byte[] array)
+            {
+                SetMemory(array);
+            }
+            else
+            {
+                Debug.Fail("Unexpected memoryOwner");
+            }
+        }
+
         public void SetMemory(IMemoryOwner<byte> memoryOwner)
         {
             _memoryOwner = memoryOwner;
 
-            AvailableMemory = _memoryOwner.Memory;
+            SetUnownedMemory(memoryOwner.Memory);
+        }
+
+        public void SetMemory(byte[] arrayPoolBuffer)
+        {
+            _memoryOwner = arrayPoolBuffer;
 
+            SetUnownedMemory(arrayPoolBuffer);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void SetUnownedMemory(Memory<byte> memory)
```

Comment:
private.

Reply:
When this type moves to System.IO.Pipelines, this will be public so I rather not have them diverge too much.